relay-server/src/actors/server.rs

```diff
@@ -1,143 +1,231 @@
-use std::net::{SocketAddr, TcpListener};
+use std::io;
+use std::net::{SocketAddr, TcpListener as StdTcpListener};
 use std::sync::Arc;
 use std::time::Duration;
 
+use axum::extract::ConnectInfo;
 use axum::http::{header, HeaderName, HeaderValue};
-use axum::ServiceExt;
-use axum_server::{AddrIncomingConfig, Handle, HttpConfig};
+use hyper_util::rt::{TokioExecutor, TokioIo, TokioTimer};
+use hyper_util::server::conn::auto::Builder;
+use hyper_util::service::TowerToHyperService;
 use relay_config::Config;
 use relay_log::tower::{NewSentryLayer, SentryHttpLayer};
-use relay_system::{Controller, Service, Shutdown};
-use tower::ServiceBuilder;
+use relay_system::{Controller, Service};
+use tokio::net::{TcpListener, TcpStream};
+use tokio::sync::{watch, Notify};
+use tower::{Layer, ServiceBuilder};
 use tower_http::set_header::SetResponseHeaderLayer;
 
 use crate::constants;
-use crate::middlewares::{
-    self, CatchPanicLayer, HandleErrorLayer, NormalizePathLayer, RequestDecompressionLayer,
-};
+use crate::middlewares::{self, CatchPanicLayer, NormalizePath, RequestDecompressionLayer};
 use crate::service::ServiceState;
 use crate::statsd::RelayCounters;
 
+/// Set a timeout for reading client request headers. If a client does not transmit the entire
+/// header within this time, the connection is closed.
+const CLIENT_HEADER_TIMEOUT: Duration = Duration::from_secs(5);
+
 /// Indicates the type of failure of the server.
 #[allow(clippy::enum_variant_names)]
-#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, thiserror::Error)]
+#[derive(Debug, thiserror::Error)]
 pub enum ServerError {
+    /// Binding failed.
+    #[error("bind to interface failed")]
+    BindFailed(#[from] io::Error),
+
     /// TLS support was not compiled in.
     #[error("SSL is no longer supported by Relay, please use a proxy in front")]
     TlsNotSupported,
 }
 
+fn listen(config: &Config) -> Result<TcpListener, ServerError> {
+    // Inform the user about a removed feature.
+    if config.tls_listen_addr().is_some()
+        || config.tls_identity_password().is_some()
+        || config.tls_identity_path().is_some()
+    {
+        return Err(ServerError::TlsNotSupported);
+    }
+
+    let listener = StdTcpListener::bind(config.listen_addr())?;
+    listener.set_nonblocking(true)?;
+    Ok(listener.try_into()?)
+}
+
+type App = NormalizePath<axum::Router>;
+
+fn make_app(service: ServiceState) -> App {
+    // Build the router middleware into a single service which runs _after_ routing. Service
+    // builder order defines layers added first will be called first. This means:
+    //  - Requests go from top to bottom
+    //  - Responses go from bottom to top
```

Comment:
Maybe we can move this to the function docs instead?

Reply:
Can do, does it make more sense for you there? The order of the middlewares is something that's important to the `ServiceBuilder`, hence the placement within the function body.

Reply:
For me yes. 
Since the entire function is actually dedicated to putting app together and the docs will do just good if they will be on the function level.

Reply:
As I was moving it and adding back an old comment that got lost, I looked at it again but it simply doesn't make sense to me. This code comment is relevant only for the implementation of the function, as it clarifies on two specifics:
1. In which order the middlewares are registered and why
2. Why some middlewares wrap the router and some others do not. 

To a user of this function, none of these matter. In fact, the user of this function does not even see that router middlewares are applied.